File: apps/api/models/emotion.py
"""Pydantic models for the Emotion Engine.

Spec reference: 06_EMOTION_ENGINE_SPEC.md
"""

from typing import Dict, List, Optional
from pydantic import BaseModel, Field


class EmotionClassifyInput(BaseModel):
    text: str = Field(..., min_length=3, description="Freeform memory, mood, or concept text.")
    occasion: Optional[str] = None
    color_preferences: Optional[str] = None


class EmotionOutput(BaseModel):
    """Structured emotion interpretation returned by the Emotion Engine."""

    primary_tone: str
    secondary_tones: List[str]
    valence: float = Field(..., ge=-1.0, le=1.0)
    arousal: float = Field(..., ge=-1.0, le=1.0)
    quadrant: str  # "Cozy & Calm" | "Bright & Joyful" | "Deep & Contemplative" | "Bold & Expressive"
    emotion_vector: Dict[str, float]  # e.g. {"calm": 0.8, "nostalgia": 0.7, "joy": 0.2}
    palette_direction: str
    density_bias: str
    movement_bias: str
    formula_candidates: List[str]  # e.g. ["crescent_asymmetrical", "arc_loose"]
    narrative_summary: str

File: packages/shared/src/types/emotion.ts
/**
 * Emotion Engine types — EC_CANON_V1 emotion schema.
 *
 * Spec reference: 06_EMOTION_ENGINE_SPEC.md
 */

export enum EmotionTone {
  JOY = "joy",
  GRIEF = "grief",
  LOVE = "love",
  NOSTALGIA = "nostalgia",
  CALM = "calm",
  CELEBRATION = "celebration",
  MELANCHOLY = "melancholy",
  REVERENCE = "reverence",
}

export type EmotionQuadrant =
  | "Cozy & Calm"
  | "Bright & Joyful"
  | "Deep & Contemplative"
  | "Bold & Expressive";

/** Sprint 1 compatible; optional fields are populated by the Emotion Engine. */
export interface EmotionInterpretation {
  tone: EmotionTone;
  intensity: number; // 0.0 – 1.0
  keywords: string[];
  summary: string;
  // Sprint 3 extended fields
  primary_tone?: string;
  secondary_tones?: string[];
  valence?: number; // -1.0 to 1.0
  arousal?: number; // -1.0 to 1.0
  quadrant?: EmotionQuadrant;
  emotion_vector?: Record<string, number>;
  palette_direction?: string;
  density_bias?: string;
  movement_bias?: string;
  narrative_summary?: string;
}

/** Input to POST /api/v1/emotion/classify */
export interface EmotionClassifyInput {
  text: string;
  occasion?: string;
  color_preferences?: string;
}

/** Full structured output from the Emotion Engine. */
export interface EmotionOutput {
  primary_tone: string;
  secondary_tones: string[];
  valence: number;
  arousal: number;
  quadrant: EmotionQuadrant;
  emotion_vector: Record<string, number>;
  palette_direction: string;
  density_bias: string;
  movement_bias: string;
  formula_candidates: string[];
  narrative_summary: string;
}
